src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs

```diff
@@ -73,13 +78,15 @@ public override void Initialize(AnalysisContext context)
 
                     foreach (var operationRoot in operationBlockContext.OperationBlocks)
                     {
-                        static bool ShouldAnalyze(IOperation op) =>
-                                (op as IBinaryOperation)?.IsComparisonOperator() == true ||
+                        bool ShouldAnalyze(IOperation op) =>
+                                ((op as IBinaryOperation)?.IsComparisonOperator() == true ||
                                 op is IInvocationOperation { TargetMethod.ReturnType.SpecialType: SpecialType.System_Boolean } ||
                                 op.Kind == OperationKind.Coalesce ||
                                 op.Kind == OperationKind.ConditionalAccess ||
                                 op.Kind == OperationKind.IsNull ||
-                                op.Kind == OperationKind.IsPattern;
+                                op.Kind == OperationKind.IsPattern)
+                                && (op.Parent is not IArgumentOperation { Parent: IInvocationOperation invocation } ||
+                                    !debugAssertMethods.Contains(invocation.TargetMethod, SymbolEqualityComparer.Default));
```

Comment:
Will this work if Debug.Assert expressions is more complex, say `Debug.Assert(MyConstant == 16 && MyConstant2 == 17);` where `MyConstant2` is another constant with value `17`?

I think a better approach would be to revert the changes to `ShouldAnalyze` and instead add a `ShouldReport(IOperation op)` helper that gets called just prior to `context.ReportDiagnostic` calls in this analyzer, and that one can do a walk up the operation tree and check if any of the ancestor `IInvocationOperation` is a call to a debug assert method.

Reply:
I thought about that approach as well, but was hesitant about implementing a plain walk-up, since we could have something like this, which IMO should be flagged: `Debug.Assert(MethodCall(MyConstant == 16));`

Reply:
Sounds reasonable. We can revisit this in future if needed. Thanks for the fix.